File: new/top_k_frequent_elements_bucket_sort.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // create a frequency map
        // create a list of n + 1 buckets (array) representing frequency. Each bucket represents potential frequency. 
        // why n + 1, because the most a number can be frequent in the given nums is nums. 
        // meaning: [1, 1, 1, 1] this is len 4, and 1 can be atmost 4 frequent. [[] [] [] [1]]
        // iterate in reverse order and get k elements from the bucket and put it in a result list and return the list
        unordered_map<int, int> freq;
        vector<vector<int>> buckets(nums.size() + 1);

        for (int num: nums) {
            freq[num]++;
        }

        for (const auto &pair: freq) {
            buckets[pair.second].push_back(pair.first);
        }

        vector<int> result;
        for (int i = buckets.size() - 1; i > 0; i--) {
            for (int num: buckets[i]) {
                result.push_back(num);
                if (result.size() == k) {
                    return result;
                }
            }
        }

        return result;
    }

    // time: O(n)
    // space: O(n)
};

File: new/python/top_k_frequent_elements_bucket_sort.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # make a frequency map
        # create array of n + 1 buckets (empty arrays) to represent frequency. Each empty bucket represent
        # frequency. Why n + 1? the worst a number can be n times frequent n being size of given list.
        # iterate from back of the bucket (most frequent) to the front (least frequent), in each of the iteration
        # iterate through each bucket and add numbers from each bucket to result array until it reaches size k, if
        # it reaches size k return the result array
        freq = dict()
        buckets = [[] for i in range(len(nums) + 1)]

        for num in nums:
            freq[num] = 1 + freq.get(num, 0)
        for val, fq in freq.items():
            buckets[fq].append(val)

        result = []
        for i in range(len(buckets) - 1, 0, -1):  # iterate in reverse order
            for num in buckets[i]:
                result.append(num)
                if len(result) == k:
                    return result

        return result
        # time: O(n)
        # space: O(n)
